pkg/codegen/schema/schema.go

```diff
@@ -0,0 +1,898 @@
+	// DefaultValue is the default value for the property, if any.
+	DefaultValue *DefaultValue
+	// IsRequired is true if the property must always be populated.
+	IsRequired bool
+	// Language specifies additional language-specific data about the property.
+	Language map[string]json.RawMessage
+}
+
+// Alias describes an alias for a Pulumi resource.
+type Alias struct {
+	// Name is the "name" portion of the alias, if any.
+	Name *string
+	// Project is the "project" portion of the alias, if any.
+	Project *string
+	// Type is the "type" portion of the alias, if any.
+	Type *string
+}
+
+// Resource describes a Pulumi resource.
+type Resource struct {
+	// Token is the resource's Pulumi type token.
+	Token string
+	// Comment is the description of the resource, if any.
+	Comment string
+	// IsProvider is true if the resource is a provider resource.
+	IsProvider bool
+	// InputProperties is the list of the resource's input properties.
+	InputProperties []*Property
+	// Properties is the list of the resource's output properties. This should be a superset of the input properties.
+	Properties []*Property
+	// StateInputs is the set of inputs used to get an existing resource, if any.
+	StateInputs *ObjectType
+	// Aliases is the list of aliases for the resource.
+	Aliases []*Alias
+	// DeprecationMessage indicates whether or not the resource is deprecated.
+	DeprecationMessage string
+	// Language specifies additional language-specific data about the resource.
+	Language map[string]json.RawMessage
+}
+
+// Function describes a Pulumi function.
+type Function struct {
+	// Token is the function's Pulumi type token.
+	Token string
+	// Comment is the description of the function, if any.
+	Comment string
+	// Inputs is the bag of input values for the function, if any.
+	Inputs *ObjectType
+	// Outputs is the bag of output values for the function, if any.
+	Outputs *ObjectType
+	// DeprecationMessage indicates whether or not the function is deprecated.
+	DeprecationMessage string
+	// Language specifies additional language-specific data about the function.
+	Language map[string]json.RawMessage
+}
+
+// Package describes a Pulumi package.
+type Package struct {
+	moduleFormat *regexp.Regexp
+
+	// Name is the unqualified name of the package (e.g. "aws", "azure", "gcp", "kubernetes". "random")
+	Name string
+	// Version is the version of the package.
+	Version *semver.Version
+	// Description is the description of the package.
+	Description string
+	// Keywords is the list of keywords that are associated with the package, if any.
+	Keywords []string
+	// Homepage is the package's homepage.
+	Homepage string
+	// License indicates which license is used for the package's contents.
+	License string
+	// Repository is the URL at which the source for the package can be found.
+	Repository string
+	// LogoURL is the URL for the package's logo, if any.
+	LogoURL string
+
+	// Types is the list of non-resource types defined by the package.
+	Types []Type
```

Comment:
might be worthwhile to explain (somewhere) how the flat list of types/resources works with any sort of namespacing concept.

Reply:
Sure thing--I'll add a comment re: the token-based approach.